src/gc/softwarewritewatch.cpp

```diff
@@ -0,0 +1,243 @@
+    assert(ALIGN_DOWN(block, sizeof(size_t)) == block);
+    assert(firstPageAddressInBlock == reinterpret_cast<uint8_t *>(GetPageAddress(block - GetTable())));
+    assert(startByteIndex < endByteIndex);
+    assert(endByteIndex <= sizeof(size_t));
+    assert(dirtyPages != nullptr);
+    assert(dirtyPageIndexRef != nullptr);
+
+    size_t &dirtyPageIndex = *dirtyPageIndexRef;
+    assert(dirtyPageIndex < dirtyPageCount);
+
+    size_t dirtyBytes = *reinterpret_cast<size_t *>(block);
+    if (dirtyBytes == 0)
+    {
+        return true;
+    }
+
+    if (startByteIndex != 0)
+    {
+        size_t numLowBitsToClear = startByteIndex * 8;
+        dirtyBytes >>= numLowBitsToClear;
+        dirtyBytes <<= numLowBitsToClear;
+    }
+    if (endByteIndex != sizeof(size_t))
+    {
+        size_t numHighBitsToClear = (sizeof(size_t) - endByteIndex) * 8;
+        dirtyBytes <<= numHighBitsToClear;
+        dirtyBytes >>= numHighBitsToClear;
+    }
+
+    while (dirtyBytes != 0)
+    {
+        DWORD bitIndex;
+        static_assert_no_msg(sizeof(size_t) <= 8);
+        if (sizeof(size_t) == 8)
+        {
+            BitScanForward64(&bitIndex, static_cast<DWORD64>(dirtyBytes));
+        }
+        else
+        {
+            BitScanForward(&bitIndex, static_cast<DWORD>(dirtyBytes));
+        }
+
+        // Each byte is only ever set to 0 or 0xff
+        assert(bitIndex % 8 == 0);
+        size_t byteMask = static_cast<size_t>(0xff) << bitIndex;
+        assert((dirtyBytes & byteMask) == byteMask);
+        dirtyBytes ^= byteMask;
+
+        DWORD byteIndex = bitIndex / 8;
+        if (clearDirty)
+        {
+            // Clear only the bytes for which pages are recorded as dirty
+            block[byteIndex] = 0;
+        }
+
+        void *pageAddress = firstPageAddressInBlock + byteIndex * OS_PAGE_SIZE;
+        assert(pageAddress >= GetHeapStartAddress());
+        assert(pageAddress < GetHeapEndAddress());
+        assert(dirtyPageIndex < dirtyPageCount);
+        dirtyPages[dirtyPageIndex] = pageAddress;
+        ++dirtyPageIndex;
+        if (dirtyPageIndex == dirtyPageCount)
+        {
+            return false;
+        }
+    }
+    return true;
+}
+
+void SoftwareWriteWatch::GetDirty(
+    void *baseAddress,
+    size_t regionByteSize,
+    void **dirtyPages,
+    size_t *dirtyPageCountRef,
+    bool clearDirty,
+    bool isRuntimeSuspended)
+{
+    VerifyCreated();
+    VerifyMemoryRegion(baseAddress, regionByteSize);
+    assert(dirtyPages != nullptr);
+    assert(dirtyPageCountRef != nullptr);
+
+    size_t dirtyPageCount = *dirtyPageCountRef;
+    if (dirtyPageCount == 0)
+    {
+        return;
+    }
+
+    if (!isRuntimeSuspended)
+    {
+        // When a page is marked as dirty, a memory barrier is not issued after the write most of the time. Issue a memory
+        // barrier on all active threads of the process now to make recent changes to dirty state visible to this thread.
+        FlushProcessWriteBuffers();
```

Comment:
Please use GCToOSInterface::FlushProcessWriteBuffers to keep the GC code OS independent